contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,447 @@
+controllers may use a combination of both approaches to mitigate the 
+deficiencies of each.
+
+### Version Equivalence
+When the target Object state of a specification type Object is revised, we wish 
+to minimize the number of mutations to generated Objects as the controller seeks
+to conform the system to its target state. That is, if a generated Object 
+already conforms to the revised target Object state, it is imperative that we 
+do not mutate it. 
+
+Failure to implement this correctly could result in the simultaneous rolling 
+restart of every Pod in every StatefulSet and DaemonSet in the system when 
+additions are made to PodTemplateSpec during a master upgrade. It is therefore 
+necessary to determine if the current target Object state is equivalent to a 
+prior version. 
+
+Since we [track the version of](#version-tracking) of generated Objects, this 
+reduces to deciding if the version of the target Object state associated with 
+the generated Object is equivalent to the current target Object state.
+Even though  [hashing](#hashing) is used to generate the `.Name` of the 
+AnyStates used to encapsulate versions of the target Object state, as we do 
+not require cryptographically strong collision resistance, and given we use a 
+[collision resolution](#collision-resolution) technique, we can't use the 
+[generated names](#unique-name-generation) of AnyStates to decide equality.
+
+We propose that two AnyStates can be considered equal if their `.Data` is 
+equivalent, but that it is not sufficient to compare the serialized 
+representation of the their `.Data`. Consider that the addition of new fields 
+to the Objects that represent the target Object state may cause the serialized 
+representation of those Objects to be unequal even when they are semantically 
+equivalent. We propose two methods that controllers may use to decide the 
+equivalence of two versions of their target Object state.
+
+1. The controller may deserialize the values of the AnyStates representing their 
+target Object state and perform a deep, semantic equality test. Here all 
+differences that do not constitute a mutation to the target Object state are 
+disregarded during the equivalence test.
+    - As it is imperative, this method may be brittle and hard to maintain.
+1. Alternatively, controllers may use a strategic merge patch to ignore fields 
+whose deletion, addition, or mutation does not constitute a revision.
+    - This approach requires all values to be JSON encodings of Objects, but the 
+    declarative nature will likely make the code more maintainable and less 
+    error prone (The business logic for the patch and comparison only needs to 
+    be implemented once).
+    
+Note that, as both methods provide a test for deep, semantic equivalence, 
+controllers that implement one method are not bound to use the same method 
+in the future. 
+    
+### Target Object State Reconciliation
```

Comment:
You need to re lable your product to match the specification that it is equivalent to, update the product in place, or recall and redistribute the product.